nn/conformer.py

```diff
@@ -0,0 +1,204 @@
+
+        self.ffn1 = _PositionwiseFeedForward(
+            d_model=enc_key_dim, d_ff=ff_dim, dropout=dropout, activation=ff_act, l2=l2)
+
+        self.ffn2 = _PositionwiseFeedForward(
+            d_model=enc_key_dim, d_ff=ff_dim, dropout=dropout, activation=ff_act, l2=l2)
+
+        self.conv_module = _ConformerConvBlock(d_model=enc_key_dim, kernel_size=conv_kernel_size)
+
+        self.mhsa_module = MultiheadAttention(d_model, att_n_heads, dropout=att_dropout)  # TODO: to be implemented
+
+    def forward(self, inp: LayerRef) -> LayerRef:
+        # FFN
+        x_ffn1_ln = layer_norm(inp)
+        x_ffn1 = self.ffn1(x_ffn1_ln)
+        x_ffn1_out = 0.5 * dropout(x_ffn1, dropout=self.dropout) + inp
+
+        # MHSA
+        x_mhsa_ln = layer_norm(x_ffn1_out)
+        x_mhsa = self.mhsa_module(x_mhsa_ln)
+        x_mhsa_out = x_mhsa + x_ffn1_out
+
+        # Conv
+        x_conv_ln = layer_norm(x_mhsa_out)
+        x_conv = self.conv_module(x_conv_ln)
+        x_conv_out = dropout(x_conv, dropout=self.dropout) + x_mhsa_out
+
+        # FFN
+        x_ffn2_ln = layer_norm(x_conv_out)
+        x_ffn2 = self.ffn2(x_ffn2_ln)
+        x_ffn2_out = 0.5 * dropout(x_ffn2, dropout=self.dropout) + x_conv_out
+
+        # last LN layer
+        return layer_norm(x_ffn2_out)
+
+
+class ConformerEncoder(Module):
+    """
+    Represents Conformer encoder architecture
+    """
+
+    def __init__(self, encoder_layer: Module, num_blocks: int, conv_kernel_size: Tuple[int, ...] = (32,), ff_act=swish,
+                 ff_dim: int = 512, dropout: float = 0.1, att_dropout: float = 0.1, enc_key_dim: int = 256,
+                 att_n_heads: int = 4, l2: float = 0.0):
+        """
+        :param encoder_layer:
+        :param num_blocks:
+        :param conv_kernel_size:
+        :param ff_act:
+        :param ff_dim:
+        :param dropout:
+        :param att_dropout:
+        :param enc_key_dim:
+        :param att_n_heads:
+        :param l2:
+        """
+        super().__init__()
+
+        self.dropout = dropout
+
+        self.conv_subsample_layer = _ConformerConvSubsampleLayer(
+            filter_sizes=[(3, 3), (3, 3)], pool_sizes=[(2, 2), (2, 2)], channel_sizes=[enc_key_dim, enc_key_dim],
+            l2=l2, dropout=dropout)
+
+        self.linear = Linear(n_out=enc_key_dim, l2=l2, with_bias=False)
+
+        self.conformer_blocks = ModuleList([
+            encoder_layer(
+                conv_kernel_size=conv_kernel_size, ff_act=ff_act, ff_dim=ff_dim, dropout=dropout,
+                att_dropout=att_dropout, enc_key_dim=enc_key_dim, att_n_heads=att_n_heads, l2=l2
+            )
+            for _ in range(num_blocks)
+        ])
+
+    def forward(self, inp: LayerRef) -> LayerRef:
+        x_subsample = self.conv_subsample_layer(inp)
+        x_linear = self.linear(x_subsample)
+        x = dropout(x_linear, dropout=self.dropout)
+        for conformer_block in self.conformer_blocks:
+            x = conformer_block(x)
```

Comment:
When you make it `Sequential` above, then this is simply:
```suggestion
        x = self.conformer_blocks(x)
```